library/src/image.cxx

```diff
@@ -121,63 +119,85 @@ image& image::operator=(image&& img) noexcept
 //----------------------------------------------------------------------------
 unsigned int image::getWidth() const
 {
-  return this->Internals->Width;
+  int dims[3];
+  this->Internals->Image->GetDimensions(dims);
+  return dims[0];
 }
 
 //----------------------------------------------------------------------------
 unsigned int image::getHeight() const
 {
-  return this->Internals->Height;
+  int dims[3];
+  this->Internals->Image->GetDimensions(dims);
+  return dims[1];
 }
 
+#ifndef F3D_NO_DEPRECATED
 //----------------------------------------------------------------------------
 image& image::setResolution(unsigned int width, unsigned int height)
 {
-  this->Internals->Width = width;
-  this->Internals->Height = height;
-  this->Internals->UpdateBufferSize();
+  this->Internals->Image->SetDimensions(width, height, 1);
+  this->Internals->Image->AllocateScalars(VTK_UNSIGNED_CHAR, this->getChannelCount());
   return *this;
 }
+#endif
 
 //----------------------------------------------------------------------------
 unsigned int image::getChannelCount() const
 {
-  return this->Internals->Channels;
+  return this->Internals->Image->GetNumberOfScalarComponents();
 }
 
+#ifndef F3D_NO_DEPRECATED
 //----------------------------------------------------------------------------
 image& image::setChannelCount(unsigned int dim)
 {
-  this->Internals->Channels = dim;
-  this->Internals->UpdateBufferSize();
+  this->Internals->Image->AllocateScalars(VTK_UNSIGNED_CHAR, dim);
   return *this;
 }
+#endif
+
+//----------------------------------------------------------------------------
+image::ChannelType image::getChannelType() const
+{
+  switch (this->Internals->Image->GetScalarType())
+  {
+    case VTK_UNSIGNED_CHAR:
+      return ChannelType::BYTE;
+    case VTK_UNSIGNED_SHORT:
+      return ChannelType::SHORT;
+    case VTK_FLOAT:
+      return ChannelType::FLOAT;
+  }
+  throw read_exception("Unknown channel type");
+}
 
 //----------------------------------------------------------------------------
 image& image::setData(unsigned char* buffer)
 {
-  std::copy(buffer, buffer + this->Internals->Buffer.size(), this->Internals->Buffer.begin());
+  int scalarSize = this->Internals->Image->GetScalarSize();
+  int totalSize = this->getWidth() * this->getHeight() * this->getChannelCount() * scalarSize;
+  unsigned char* internalBuffer =
+    reinterpret_cast<unsigned char*>(this->Internals->Image->GetScalarPointer());
+  std::copy(buffer, buffer + totalSize, internalBuffer);
```

Comment:
check for ChannelType ?

Reply:
there's no way to know what's inside the user buffer